monai/networks/layers/conjugate_gradient.py

```diff
@@ -0,0 +1,120 @@
+# Copyright (c) MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from typing import Callable, Optional
+
+import torch
+from torch import nn
+
+
+class ConjugateGradient(nn.Module):
+    """
+    Congugate Gradient (CG) solver for linear systems Ax = y.
+
+    For A (linear_op) that is positive definite and self-adjoint, CG is
+    guaranteed to converge CG is often used to solve linear systems of the form
+    Ax = y, where A is too large to store explicitly, but can be computed via a
+    linear operator.
+
+    As a result, here we won't set A explicitly as a matrix, but rather as a
+    linear operator. For example, A could be a FFT/IFFT operation
+    """
+
+    def __init__(self, linear_op: Callable, num_iter: int, dbprint: Optional[bool] = False):
+        """
+        Args:
+            linear_op: Linear operator
+            num_iter: Number of iterations to run CG
+            dbprint [False]: Print residual at each iteration
+        """
+        super().__init__()
+
+        self.A = linear_op
+        self.num_iter = num_iter
+        self.dbprint = dbprint
+
+    def _zdot(self, x1: torch.Tensor, x2: torch.Tensor) -> torch.Tensor:
+        """
+        Complex dot product between tensors x1 and x2: sum(x1.*x2)
+        """
+        if torch.is_complex(x1):
+            assert torch.is_complex(x2), "x1 and x2 must both be complex"
+            return torch.sum(x1.conj() * x2)
+        else:
+            return torch.sum(x1 * x2)
+
+    def _zdot_single(self, x: torch.Tensor) -> torch.Tensor:
+        """
+        Complex dot product between tensor x and itself
+        """
+        res = self._zdot(x, x)
+        if torch.is_complex(res):
+            return res.real
+        else:
+            return res
```

Comment:
These two methods could be utility functions placed elsewhere since they don't rely on `self` internally. Moving them out would reduce the size of this class.
```suggestion
    def _zdot(self, x1: torch.Tensor, x2: torch.Tensor) -> torch.Tensor:
        """
        Complex dot product between tensors x1 and x2: sum(x1.*x2)
        """
        if torch.is_complex(x1):
            if not torch.is_complex(x2):
                raise ValueError("x2 must be complex if x1 is complex")
            x1=x1.conj()
        
        return torch.sum(x1 * x2)

    def _zdot_single(self, x: torch.Tensor) -> torch.Tensor:
        """
        Complex dot product between tensor x and itself
        """
        res = self._zdot(x, x)
        return res.real if torch.is_complex(res) else res
```